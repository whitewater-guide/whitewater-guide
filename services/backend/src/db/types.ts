export interface RawTimestamped {
  created_at: Date;
  updated_at: Date;
}

export interface WithLanguage {
  language: string;
}

export const timestampResolvers = {
  createdAt: (node: RawTimestamped) => new Date(node.created_at).toISOString(),
  updatedAt: (node: RawTimestamped) => new Date(node.updated_at).toISOString(),
};
