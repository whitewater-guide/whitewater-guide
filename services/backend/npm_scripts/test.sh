#!/usr/bin/env bash

set -o allexport
source ../../config/.env.test
source .env.test
set +o allexport

./node_modules/.bin/jest -i --forceExit "$@"
