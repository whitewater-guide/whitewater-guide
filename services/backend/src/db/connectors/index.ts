export * from './types';
export * from './OffsetConnector';
export * from './createConnectors';
