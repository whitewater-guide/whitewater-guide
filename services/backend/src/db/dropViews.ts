import Knex from 'knex';

export async function dropViews(db: Knex, ...views: string[]) {
  for (const view of views) {
    await db.raw(`DROP VIEW IF EXISTS ${view}_view`);
  }
}
