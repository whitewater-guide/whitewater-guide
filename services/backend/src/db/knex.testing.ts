import knex from 'knex';

const test: knex.Config = {
  client: 'pg',
  connection: {
    host: process.env.POSTGRES_HOST!,
    port: 5432,
    database: process.env.POSTGRES_DB!,
    user: 'postgres',
    password: process.env.POSTGRES_PASSWORD!,
  },
  pool: {
    min: 0,
    max: 1,
    idleTimeoutMillis: 500,
  },
  migrations: {
    loadExtensions: ['.js'],
    tableName: 'migrations',
    directory: './dist/migrations',
  } as any,
  seeds: {
    loadExtensions: ['.js'],
    directory: './dist/seeds/test',
  } as any,
  debug: false, // process.env.DATABASE_DEBUG === 'true',
};

module.exports = test;
