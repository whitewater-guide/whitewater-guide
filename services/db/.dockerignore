.env.*
Dockerfile
package.json
/config/x-dev-restore.sh
/config/*.bak
/scripts
/.env
