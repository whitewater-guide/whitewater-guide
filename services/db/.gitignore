.env
!.env.*.secret
/config/*.bak
/config/*.csv
