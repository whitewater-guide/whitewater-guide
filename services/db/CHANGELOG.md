## 0.0.25 (2019-01-31)
