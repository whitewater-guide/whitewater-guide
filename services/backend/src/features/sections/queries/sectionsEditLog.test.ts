import { holdTransaction, rollbackTransaction } from '~/db';
import {
  ADMIN,
  ADMIN_ID,
  EDITOR_GA_EC,
  TEST_USER,
} from '~/seeds/test/01_users';
import { REGION_GEORGIA } from '~/seeds/test/04_regions';
import {
  SECTION_EDIT_LOG_ENTRY_1,
  SECTION_EDIT_LOG_ENTRY_2,
  SECTION_EDIT_LOG_ENTRY_3,
  SECTION_EDIT_LOG_ENTRY_4,
} from '~/seeds/test/15_sections_edit_log';
import { anonContext, fakeContext, runQuery } from '~/test';
import { ApolloErrorCodes } from '@whitewater-guide/commons';

beforeEach(holdTransaction);
afterEach(rollbackTransaction);

const query = `
query sectionsEditLog($filter: SectionsEditLogFilter, $page: Page) {
  sectionsEditLog(filter: $filter, page: $page) {
    nodes {
      __typename
      id
      section {
        __typename
        id
        name
        river {
          __typename
          id
          name
        }
        region {
          __typename
          id
          name
        }
      }
      editor {
        __typename
        id
        name
      }
      action
      diff
      createdAt
    }
    count
  }
}
`;

describe('resolvers chain', () => {
  it('anon should not pass', async () => {
    const result = await runQuery(query, {}, anonContext());
    expect(result).toHaveGraphqlError(ApolloErrorCodes.UNAUTHENTICATED);
  });

  it('user should not pass', async () => {
    const result = await runQuery(query, {}, fakeContext(TEST_USER));
    expect(result).toHaveGraphqlError(ApolloErrorCodes.FORBIDDEN);
  });

  it('editor should not pass', async () => {
    const result = await runQuery(query, {}, fakeContext(EDITOR_GA_EC));
    expect(result).toHaveGraphqlError(ApolloErrorCodes.FORBIDDEN);
  });
});

describe('data', () => {
  it('should return all log entries', async () => {
    const result = await runQuery(query, {}, fakeContext(ADMIN));
    expect(result.errors).toBeUndefined();
    expect(result.data!.sectionsEditLog).toMatchSnapshot();
  });

  it('should paginate', async () => {
    const result = await runQuery(
      query,
      { page: { limit: 1, offset: 1 } },
      fakeContext(ADMIN),
    );
    expect(result.errors).toBeUndefined();
    expect(result.data!.sectionsEditLog).toMatchObject({
      count: 4,
      nodes: [
        {
          id: SECTION_EDIT_LOG_ENTRY_3,
        },
      ],
    });
  });

  it('should filter by region', async () => {
    const result = await runQuery(
      query,
      { filter: { regionId: REGION_GEORGIA } },
      fakeContext(ADMIN),
    );
    expect(result.errors).toBeUndefined();
    expect(result.data!.sectionsEditLog).toMatchObject({
      count: 3,
      nodes: [
        { id: SECTION_EDIT_LOG_ENTRY_3 },
        { id: SECTION_EDIT_LOG_ENTRY_2 },
        { id: SECTION_EDIT_LOG_ENTRY_1 },
      ],
    });
  });

  it('should filter by editor', async () => {
    const result = await runQuery(
      query,
      { filter: { editorId: ADMIN_ID } },
      fakeContext(ADMIN),
    );
    expect(result.errors).toBeUndefined();
    expect(result.data!.sectionsEditLog).toMatchObject({
      count: 2,
      nodes: [
        { id: SECTION_EDIT_LOG_ENTRY_4 },
        { id: SECTION_EDIT_LOG_ENTRY_3 },
      ],
    });
  });

  it('should filter by region AND editor', async () => {
    const result = await runQuery(
      query,
      { filter: { regionId: REGION_GEORGIA, editorId: ADMIN_ID } },
      fakeContext(ADMIN),
    );
    expect(result.errors).toBeUndefined();
    expect(result.data!.sectionsEditLog).toMatchObject({
      count: 1,
      nodes: [{ id: SECTION_EDIT_LOG_ENTRY_3 }],
    });
  });
});
