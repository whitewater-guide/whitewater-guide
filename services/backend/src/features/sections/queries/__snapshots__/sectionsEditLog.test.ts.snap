// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`data should return all log entries 1`] = `
Array [
  Object {
    "__typename": "SectionEditLogEntry",
    "action": "update",
    "createdAt": "2017-02-04T00:00:00.000Z",
    "editor": Object {
      "__typename": "User",
      "id": "bed59990-749d-11e7-8cf7-a6006ad3dba0",
      "name": "Ivan Ivanov",
    },
    "id": "b4d42c34-5a2d-4136-b11e-35fda268f080",
    "section": Object {
      "__typename": "Section",
      "id": "21f2351e-d52a-11e7-9296-cec278b6b50a",
      "name": "Amot",
      "region": Object {
        "__typename": "Region",
        "id": "b968e2b2-76c5-11e7-b5a5-be2e44b06b34",
        "name": "Norway",
      },
      "river": Object {
        "__typename": "River",
        "id": "d4396dac-d528-11e7-9296-cec278b6b50a",
        "name": "Sjoa",
      },
    },
  },
  Object {
    "__typename": "SectionEditLogEntry",
    "action": "update",
    "createdAt": "2017-02-03T00:00:00.000Z",
    "editor": Object {
      "__typename": "User",
      "id": "bed59990-749d-11e7-8cf7-a6006ad3dba0",
      "name": "Ivan Ivanov",
    },
    "id": "badb7ed1-c43a-4297-842a-7f12a59501a1",
    "section": Object {
      "__typename": "Section",
      "id": "f73b533c-5db4-11e8-9c2d-fa7ae01bbebc",
      "name": "Квалификация -> До шайбы",
      "region": Object {
        "__typename": "Region",
        "id": "8e119768-37f3-11e8-b467-0ed5f89f718b",
        "name": "Грузия",
      },
      "river": Object {
        "__typename": "River",
        "id": "b80554ba-5db4-11e8-9c2d-fa7ae01bbebc",
        "name": "Бжужа",
      },
    },
  },
  Object {
    "__typename": "SectionEditLogEntry",
    "action": "create",
    "createdAt": "2017-02-02T00:00:00.000Z",
    "editor": Object {
      "__typename": "User",
      "id": "42cf6da0-5dc0-11e8-9c2d-fa7ae01bbebc",
      "name": "Kaki",
    },
    "id": "23d7e919-77e1-412b-87d6-7420fe1c70b6",
    "section": Object {
      "__typename": "Section",
      "id": "e6e0e826-5db4-11e8-9c2d-fa7ae01bbebc",
      "name": "Long Race",
      "region": Object {
        "__typename": "Region",
        "id": "8e119768-37f3-11e8-b467-0ed5f89f718b",
        "name": "Georgia",
      },
      "river": Object {
        "__typename": "River",
        "id": "b80554ba-5db4-11e8-9c2d-fa7ae01bbebc",
        "name": "Bzhuzha",
      },
    },
  },
  Object {
    "__typename": "SectionEditLogEntry",
    "action": "delete",
    "createdAt": "2017-02-01T00:00:00.000Z",
    "editor": Object {
      "__typename": "User",
      "id": "42cf6da0-5dc0-11e8-9c2d-fa7ae01bbebc",
      "name": "Kaki",
    },
    "id": "9ac6562b-91fa-49f2-a765-4efe3a4d2680",
    "section": Object {
      "__typename": "Section",
      "id": "2bc05fa8-34d5-447a-bf55-34a9a7e726e1",
      "name": "deleted_section1_name",
      "region": Object {
        "__typename": "Region",
        "id": "8e119768-37f3-11e8-b467-0ed5f89f718b",
        "name": "Georgia",
      },
      "river": Object {
        "__typename": "River",
        "id": "b80554ba-5db4-11e8-9c2d-fa7ae01bbebc",
        "name": "Bzhuzha",
      },
    },
  },
]
`;
